File: pkg/msg/fs.go
package msg

import (
	"errors"
	"fmt"
	"time"
)

const fsEndpoint = 0x3

// FSInfo describes a file system entry.
type FSInfo struct {
	Name  string
	IsDir bool
	Size  uint32
}

// StatPath retrieves information about a file system entry.
func StatPath(s *Session, path string, timeout time.Duration) (*FSInfo, error) {
	// send command
	cmd := pack("os", uint8(0), path)
	err := fsSend(s, cmd, false, timeout)
	if err != nil {
		return nil, err
	}

	// await reply
	reply, err := fsReceive(s, false, timeout)
	if err != nil {
		return nil, err
	}

	// verify "info" reply
	if len(reply) == 6 && reply[0] != 1 {
		return nil, fmt.Errorf("invalid message")
	}

	// unpack "info" reply
	args := unpack("oi", reply[1:])

	return &FSInfo{
		IsDir: args[0].(uint8) == 1,
		Size:  args[1].(uint32),
	}, nil
}

// ListDir retrieves a list of file system entries in a directory.
func ListDir(s *Session, dir string, timeout time.Duration) ([]FSInfo, error) {
	// send command
	cmd := pack("os", uint8(1), dir)
	err := fsSend(s, cmd, false, timeout)
	if err != nil {
		return nil, nil
	}

	// prepare infos
	var infos []FSInfo

	for {
		// await reply
		reply, err := fsReceive(s, true, timeout)
		if errors.Is(err, Ack) {
			return infos, nil
		} else if err != nil {
			return nil, err
		}

		// verify "info" reply
		if len(reply) < 7 && reply[0] != 1 {
			return nil, fmt.Errorf("invalid message")
		}

		// unpack "info" reply
		args := unpack("ois", reply[1:])

		// add info
		infos = append(infos, FSInfo{
			Name:  args[2].(string),
			IsDir: args[0].(uint8) == 1,
			Size:  args[1].(uint32),
		})
	}
}

// ReadFile reads the contents of a file.
func ReadFile(s *Session, file string, report func(uint32), timeout time.Duration) ([]byte, error) {
	// stat file
	info, err := StatPath(s, file, timeout)
	if err != nil {
		return nil, err
	}

	// prepare buffer
	data := make([]byte, info.Size)

	// read file in chunks of 5 KB
	var offset uint32
	for offset < info.Size {
		// determine length
		length := min(5000, info.Size-offset)

		// read range
		rng, err := ReadFileRange(s, file, offset, length, func(pos uint32) {
			if report != nil {
				report(offset + pos)
			}
		}, timeout)
		if err != nil {
			return nil, err
		}

		// append range
		copy(data[offset:], rng)
		offset += uint32(len(rng))
	}

	return data, nil
}

// ReadFileRange reads a range of bytes from a file.
func ReadFileRange(s *Session, file string, offset, length uint32, report func(uint32), timeout time.Duration) ([]byte, error) {
	// send "open" command
	cmd := pack("oos", uint8(2), uint8(0), file)
	err := fsSend(s, cmd, true, timeout)
	if err != nil {
		return nil, err
	}

	// send "read" command
	cmd = pack("oii", uint8(3), offset, length)
	err = fsSend(s, cmd, false, timeout)
	if err != nil {
		return nil, err
	}

	// prepare data
	data := make([]byte, length)

	// prepare counter
	var count uint32

	for {
		// await reply
		reply, err := fsReceive(s, true, timeout)
		if errors.Is(err, Ack) {
			break
		} else if err != nil {
			return nil, err
		}

		// verify "chunk" reply
		if len(reply) < 5 && reply[0] != 2 {
			return nil, fmt.Errorf("invalid message")
		}

		// unpack offset
		replyOffset := unpack("i", reply[1:])[0].(uint32)

		// verify offset
		if replyOffset != offset+count {
			return nil, fmt.Errorf("invalid offset")
		}

		// append data
		copy(data[count:], reply[5:])

		// increment
		count += uint32(len(reply) - 5)

		// report length
		if report != nil {
			report(count)
		}
	}

	// send "close" command
	cmd = pack("o", uint8(5))
	err = fsSend(s, cmd, true, timeout)
	if err != nil {
		return nil, err
	}

	return data, nil
}

// WriteFile writes data to a file.
func WriteFile(s *Session, file string, data []byte, report func(uint32), timeout time.Duration) error {
	// send "create" command
	cmd := pack("oos", uint8(2), uint8((1<<0)|(1<<2)), file)
	err := fsSend(s, cmd, true, timeout)
	if err != nil {
		return err
	}

	// get width
	width := s.Channel().Width()

	// get MTU
	mtu, err := s.GetMTU(time.Second)
	if err != nil {
		return err
	}

	// subtract overhead
	mtu -= 6

	// write data in chunks
	num := 0
	offset := 0
	for offset < len(data) {
		// determine chunk size and chunk data
		chunkSize := min(int(mtu), len(data)-offset)
		chunkData := data[offset : offset+chunkSize]

		// determine mode
		acked := num%width == 0

		// prepare "write" command (acked or silent & sequential)
		cmd := pack("ooib", uint8(4), b2v(acked, uint8(0), uint8(1<<0|1<<1)), uint32(offset), chunkData)

		// send "write" command
		err = fsSend(s, cmd, false, 0)
		if err != nil {
			return err
		}

		// receive ack or "error" replies
		if acked {
			_, err := fsReceive(s, true, timeout)
			if err != nil && !errors.Is(err, Ack) {
				return err
			}
		}

		// increment offset
		offset += chunkSize

		// report offset
		if report != nil {
			report(uint32(offset))
		}

		// increment count
		num++
	}

	// send "close" command
	cmd = pack("o", uint8(5))
	err = fsSend(s, cmd, true, timeout)
	if err != nil {
		return err
	}

	return nil
}

// RenamePath renames a file system entry.
func RenamePath(s *Session, from, to string, timeout time.Duration) error {
	// send command
	cmd := pack("osos", uint8(6), from, uint8(0), to)
	err := fsSend(s, cmd, true, timeout)
	if err != nil {
		return err
	}

	return nil
}

// RemovePath removes a file system entry.
func RemovePath(s *Session, path string, timeout time.Duration) error {
	// send command
	cmd := pack("os", uint8(7), path)
	err := fsSend(s, cmd, true, timeout)
	if err != nil {
		return err
	}

	return nil
}

// SHA256File retrieves the SHA-256 hash of a file.
func SHA256File(s *Session, file string, timeout time.Duration) ([]byte, error) {
	// send command
	cmd := pack("os", uint8(8), file)
	err := fsSend(s, cmd, false, timeout)
	if err != nil {
		return nil, err
	}

	// await reply
	reply, err := fsReceive(s, false, timeout)
	if err != nil {
		return nil, err
	}

	// verify "hash" reply
	if len(reply) != 33 && reply[0] != 3 {
		return nil, fmt.Errorf("invalid message")
	}

	// return hash
	return reply[1:], nil
}

// MakePath creates a directory path.
func MakePath(s *Session, path string, timeout time.Duration) error {
	// send command
	cmd := pack("os", uint8(9), path)
	err := fsSend(s, cmd, true, timeout)
	if err != nil {
		return err
	}

	return nil
}

/* Helpers */

func fsReceive(s *Session, expectAck bool, timeout time.Duration) ([]byte, error) {
	// await reply
	reply, err := s.Receive(fsEndpoint, expectAck, timeout)
	if err != nil {
		return nil, err
	}

	// handle errors
	if reply[0] == 0 {
		return nil, fmt.Errorf("posix error: %s", reply[1:])
	}

	return reply, nil
}

func fsSend(s *Session, data []byte, awaitAck bool, timeout time.Duration) error {
	// send data
	if awaitAck {
		return s.Send(fsEndpoint, data, timeout)
	} else {
		return s.Send(fsEndpoint, data, 0)
	}
}

File: python/naos/fs.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Callable, List, Optional

from .session import Session
from .utils import pack, unpack

_fs_endpoint = 0x03


@dataclass
class FSInfo:
    name: str
    is_dir: bool
    size: int


async def stat_path(session: Session, path: str, timeout: float = 5.0) -> FSInfo:
    """Return information about the given path."""

    # send command
    cmd = pack("os", 0, path)
    await _send(session, cmd, False, timeout)

    # await reply
    reply = await _receive(session, False, timeout)

    # verify "info" reply
    if len(reply) != 6 or reply[0] != 1:
        raise RuntimeError("invalid reply")

    # unpack "info" reply
    is_dir, size = unpack("oi", reply[1:])

    return FSInfo(path.rsplit("/", 1)[-1], is_dir == 1, size)


async def list_dir(session: Session, dir: str, timeout: float = 5.0) -> List[FSInfo]:
    """Return a list of entries in the given directory."""

    # send command
    cmd = pack("os", 1, dir)
    await _send(session, cmd, False, timeout)

    # prepare infos
    infos = []

    while True:
        # await reply
        reply = await _receive(session, True, timeout)
        if reply is None:
            return infos

        # verify "info" reply
        if len(reply) < 7 or reply[0] != 1:
            raise RuntimeError("invalid reply")

        # unpack "info" reply
        is_dir, size, name = unpack("ois", reply[1:])

        # add info
        infos.append(FSInfo(name, is_dir == 1, size))


async def read_file(
    session: Session,
    file: str,
    report: Optional[Callable[[int], None]] = None,
    timeout: float = 5.0,
) -> bytes:
    """Read the contents of the given file."""

    # stat file
    info = await stat_path(session, file, timeout)

    # prepare data
    data = bytearray()

    # read file in chunks of 5 KB
    while len(data) < info.size:
        # determine length
        length = min(5000, info.size - len(data))

        # read range
        offset = len(data)
        range_ = await read_file_range(
            session,
            file,
            offset,
            length,
            (lambda pos: report(offset + pos)) if report else None,
            timeout,
        )

        # append range
        data += range_

    return bytes(data)


async def read_file_range(
    session: Session,
    file: str,
    offset: int,
    length: int,
    report: Optional[Callable[[int], None]] = None,
    timeout: float = 5.0,
) -> bytes:
    """Read a range of the given file."""

    # send "open" command
    cmd = pack("oos", 2, 0, file)
    await _send(session, cmd, True, timeout)

    # send "read" command
    cmd = pack("oii", 3, offset, length)
    await _send(session, cmd, False, timeout)

    # prepare data
    data = bytearray()

    while True:
        # await reply
        reply = await _receive(session, True, timeout)
        if reply is None:
            break

        # verify "chunk" reply
        if len(reply) <= 5 or reply[0] != 2:
            raise RuntimeError("invalid reply")

        # verify offset
        reply_offset = unpack("i", reply[1:5])[0]
        if reply_offset != offset + len(data):
            raise RuntimeError("invalid offset")

        # append data
        data += reply[5:]

        # report length
        if report:
            report(len(data))

    # send "close" command
    cmd = pack("o", 5)
    await _send(session, cmd, True, timeout)

    return bytes(data)


async def write_file(
    session: Session,
    file: str,
    data: bytes,
    report: Optional[Callable[[int], None]] = None,
    timeout: float = 5.0,
):
    """Write the given data to the given file."""

    # send "create" command (create & truncate)
    cmd = pack("oos", 2, (1 << 0) | (1 << 2), file)
    await _send(session, cmd, True, timeout)

    # get width
    width = session.channel().width()

    # get MTU and subtract overhead
    mtu = await session.get_mtu(timeout) - 6

    # write data in chunks
    num = 0
    offset = 0
    while offset < len(data):
        # determine chunk
        chunk = data[offset : offset + mtu]

        # determine mode
        acked = num % width == 0 or offset + len(chunk) >= len(data)

        # prepare "write" command (sequential or silent & sequential)
        cmd = pack("ooib", 4, 1 << 1 if acked else (1 << 0) | (1 << 1), offset, chunk)

        # send "write" command
        await _send(session, cmd, False, timeout)

        # receive ack or "error" replies
        if acked:
            await _receive(session, True, timeout)

        # increment offset
        offset += len(chunk)

        # report offset
        if report:
            report(offset)

        # increment count
        num += 1

    # send "close" command
    cmd = pack("o", 5)
    await _send(session, cmd, True, timeout)


async def rename_path(session: Session, from_: str, to: str, timeout: float = 5.0):
    """Rename the given path."""

    # send command
    cmd = pack("osos", 6, from_, 0, to)
    await _send(session, cmd, True, timeout)


async def remove_path(session: Session, path: str, timeout: float = 5.0):
    """Remove the given path."""

    # send command
    cmd = pack("os", 7, path)
    await _send(session, cmd, True, timeout)


async def sha256_file(session: Session, file: str, timeout: float = 5.0) -> bytes:
    """Return the SHA-256 hash of the given file."""

    # send command
    cmd = pack("os", 8, file)
    await _send(session, cmd, False, timeout)

    # await reply
    reply = await _receive(session, False, timeout)

    # verify "chunk" reply
    if len(reply) != 33 or reply[0] != 3:
        raise RuntimeError("invalid reply")

    # return hash
    return reply[1:]


async def make_path(session: Session, path: str, timeout: float = 5.0):
    """Create a directory at the given path."""

    # send command
    cmd = pack("os", 9, path)
    await _send(session, cmd, True, timeout)


# Helpers


async def _receive(
    session: Session, expect_ack: bool, timeout: float = 5.0
) -> Optional[bytes]:
    # receive reply
    data, _ = await session.receive(_fs_endpoint, expect_ack, timeout)
    if data is None:
        return None

    # handle errors
    if data[0] == 0:
        raise RuntimeError(f"posix error: {data[1]}")

    return data


async def _send(session: Session, data: bytes, await_ack: bool, timeout: float = 5.0):
    # send command
    await session.send(_fs_endpoint, data, 0)

    # await ack
    if await_ack:
        await _receive(session, True, timeout)
